pypit/core/artracewave.py

```diff
@@ -0,0 +1,1235 @@
+            ytfit = ytfit[:dx]
+            wmask = wmask[np.where(wmask < (xtfit.size+dx))]
+
+        # Perform a scanning polynomial fit to the tilts
+        wmfit = np.where(ytfit != maskval)
+        if wmfit[0].size > tilt_settings['tilts']['order'] + 1:
+            cmfit = arutils.func_fit(xtfit[wmfit], ytfit[wmfit],
+                                     tilt_settings['tilts']['function'],
+                                     tilt_settings['tilts']['order'],
+                                     minv=0.0, maxv=msarc.shape[1] - 1.0)
+            model = arutils.func_val(cmfit, xtfit, tilt_settings['tilts']['function'],
+                                     minv=0.0, maxv=msarc.shape[1] - 1.0)
+        else:
+            aduse[j] = False
+            badlines += 1
+            continue
+
+        # Can this actually happen??
+        if maskval in model:
+            # Model contains masked values
+            aduse[j] = False
+            badlines += 1
+            continue
+
+        # Perform a robust polynomial fit to the traces
+        wmsk, mcoeff = arutils.robust_polyfit(xtfit[wmask], ytfit[wmask],
+                                              tilt_settings['tilts']['order'],
+                                              function=tilt_settings['tilts']['function'],
+                                              sigma=2.0, minv=0.0, maxv=msarc.shape[1] - 1.0)
+
+        # Save model
+        model = arutils.func_val(mcoeff, xtfit, tilt_settings['tilts']['function'],
+                                 minv=0.0, maxv=msarc.shape[1] - 1.0)
+        xmodel.append(xtfit)
+        ymodel.append(model)
+
+        # Save
+        xtilt[xint:lastx, j] = xtfit / (msarc.shape[1] - 1.0)
+        # These should be un-normalized for now
+        pcen = pixcen[arcdet[j], slit]
+        ytilt[xint:lastx, j] = model[pcen-int(xtfit[wmask[0]])]
+        mtilt[xint:lastx, j] = model
+
+    # Save
+    trcdict['xmodel'] = xmodel
+    trcdict['ymodel'] = ymodel
+    trcdict["aduse"] = aduse
+
+    # Return
+    all_tilts = (xtilt, ytilt, mtilt, wtilt)
+    return badlines, all_tilts
+
+
+def new_tilts_image(tilts, lordloc, rordloc, pad, sz_y):
+    """
+    Using the tilt (assumed to be fit with a first order polynomial)
+    generate an image of the tilts for each slit.
+
+    Parameters
+    ----------
+    tilts : ndarray
+      An (m x n) 2D array specifying the tilt (i.e. gradient) at each
+      pixel along the spectral direction (m) for each slit (n).
+    lordloc : ndarray
+      Location of the left slit edges
+    rordloc : ndarray
+      Location of the right slit edges
+    pad : int
+      Set the tilts within each slit, and extend a number of pixels
+      outside the slit edges (this number is set by pad).
+    sz_y : int
+      Number of detector pixels in the spatial direction.
+
+    Returns
+    -------
+    tiltsimg : ndarray
+      An image the same size as the science frame, containing values from 0-1.
```

Comment:
Technically, the values could be a little above 1 and below 0, if there's a tilt (i.e. for the bottom row of pixels, the central pixel of the slit may be 0, but a little to the left it could be negative)